stackrox/stackrox
# ROX-15147: Add CentralCertificate and ScannerCertificate cards to System Health

## Description

### Problems

Although central and scanner certificate have banners when expiration is within 2 weeks, it is consistent to include in System Health so:

1. It is possible to know the amount of time until expiration, similar to sensor certificate expiration on Clusters page (although with different mapping from time to color) so customers can plan ahead.

2. It becomes possible to download certificate on demand so customer can update certificates on demand in response to an internal security incident.

### Analysis

Similar to src/Containers/MainPage/Banners/CredentialExpiryBanner.tsx

### Solution

1. Add **Central certificate** and **Scanner certificate** cards at the **end** of System Health page, so cards which are more likely to have errors are less likely to require scrolling.

2. Move files from Components and utils folders into more meaningful folders.

3. Factor out reusable CardHeaderIcons.

4. Factor out common code into utils/credentialExpiry.ts file.

### Residue

1. Move icon to left in header of other cards.
2. Write integration tests.
3. Factor out external link as a reusable component.
4. Delete obsolete utils files after Ivan rewrites IntegrationHealth cards.

## Checklist

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui
3. `yarn start` in ui

### Manual testing

**NOTE**: A follow up pull request will move icon to left in header of other cards.

1. Visit /main/system-health

    * See GET /v1/credentialexpiry requests.
        ![credentialExpiry](https://github.com/stackrox/stackrox/assets/11862657/c381fbfb-d49a-4559-807d-7941576c0729)

    * Temporarily edit code to see Spinner while fetching initial request.
        ![Spinner](https://github.com/stackrox/stackrox/assets/11862657/65ea88c4-92d9-490f-a5ed-299f9dea7039)

    * Temporarily edit code to see error message if GET /v1/credentialexpiry request fails.
        ![errorMessageFetching](https://github.com/stackrox/stackrox/assets/11862657/99a3c9c7-a2cd-47ac-8ce2-5c3ef61b8d48)

    * Temporarily edit code to see warning if expiration date is in 14 days.
        ![warning](https://github.com/stackrox/stackrox/assets/11862657/d6c2dfeb-66dc-4f36-9a7f-c947dbecdf2a)

    * Temporarily edit code to see danger if expiration date is in 3 days.
        ![danger](https://github.com/stackrox/stackrox/assets/11862657/cccd834b-3e0e-4e3d-ab32-0f93fe43d812)

    * Temporarily edit code to see sentence for non-administrator role if expiration date is in 3 days.
        ![danger_non-admin](https://github.com/stackrox/stackrox/assets/11862657/b6e0ad27-101e-401e-b94e-df462bcc0211)

    * Temporarily edit code to see absence of sentence for non-administrator role if expiration date is greater than 14 days.
        ![success_non-admin](https://github.com/stackrox/stackrox/assets/11862657/7c4202b2-f703-402d-bc21-9d90e811c639)

    * Temporarily edit code to see disabled and spinner for **Download YAML** button.
        ![Spinner_Download_YAML](https://github.com/stackrox/stackrox/assets/11862657/389f8435-5073-4817-9ad0-06a5a9820f7a)

    * Temporarily edit code to see error message if POST /api/extensions/certgen request fails.
        ![errorMessageDownloading](https://github.com/stackrox/stackrox/assets/11862657/29943180-2cd0-4146-8b35-426cb57b6c69)

    * Click external link to see **Reissuing internal certificates** docs page open in a new browser tab.

### Integration testing

1. `yarn cypress-open` in ui/apps/platform

    * credentialExpiry/credentialExpiry.test.js
